clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java

```diff
@@ -0,0 +1,263 @@
+ *         {@link ConsumerRebalanceListener} callback methods.
+ *     </li>
+ *     <li>
+ *         On the application thread, when the {@link PartitionReconciliationStartedEvent} is received, execute the
+ *         {@link ConsumerRebalanceListener#onPartitionsRevoked(Collection)} and
+ *         {@link ConsumerRebalanceListener#onPartitionsAssigned(Collection)} callback methods
+ *     </li>
+ *     <li>
+ *         Enqueue a corresponding {@link PartitionReconciliationCompleteEvent} so that the background thread
+ *         will know the listener was invoked and the result of the invocation
+ *     </li>
+ *     <li>
+ *         On the background thread, process the {@link PartitionReconciliationCompleteEvent}, which should call the
+ *         {@link MembershipManager#completeReconcile(Set, Set, Optional)} method. This method will call
+ *         {@link #completeReconcile(Set, Set)} to remove the revoked partitions and add the assigned partitions in the
+ *         {@link SubscriptionState#assignFromSubscribed(Collection) current assignment} and then make a note to
+ *         send a {@link ConsumerGroupHeartbeatRequest} to the group coordinator on its next pass of
+ *         {@link HeartbeatRequestManager#poll(long)}
+ *     </li>
+ * </ol>
+ *
+ * <p/>
+ *
+ * The comparison against the {@link SubscriptionState#assignedPartitions() current set of assigned partitions} and
+ * the {@link Assignment#topicPartitions() target set of assigned partitions} is performed by essentially
+ * <em>flattening</em> the respective entries into two sets of {@link TopicPartition partitions}
+ * which are then compared using basic {@link Set} comparisons.
+ */
+public class AssignmentReconciler {
+
+    private final Logger log;
+    private final SubscriptionState subscriptions;
+    private final ConsumerMetadata metadata;
+    private final BlockingQueue<BackgroundEvent> backgroundEventQueue;
+
+    AssignmentReconciler(LogContext logContext,
+                         SubscriptionState subscriptions,
+                         ConsumerMetadata metadata,
+                         BlockingQueue<BackgroundEvent> backgroundEventQueue) {
+        this.log = logContext.logger(getClass());
+        this.subscriptions = subscriptions;
+        this.metadata = metadata;
+        this.backgroundEventQueue = backgroundEventQueue;
+    }
+
+    void startLost() {
+        SortedSet<TopicPartition> partitionsToLose = getPartitionsToLose();
+
+        if (partitionsToLose.isEmpty()) {
+            log.debug("Skipping invocation of {} callbacks as no partitions changed in the new assignment",
+                    ConsumerRebalanceListener.class.getSimpleName());
+            return;
+        }
+
+        log.debug("Enqueuing event to invoke {} callbacks", ConsumerRebalanceListener.class.getSimpleName());
+        backgroundEventQueue.add(new PartitionLostStartedEvent(partitionsToLose));
```

Comment:
We could also skip this one if we don't have it.

Reply:
Yes, we initially wanted to skip this rigamarole, if possible. It's not clear to me how we know if we have a "valid" listener or not. Let me noodle on it a bit.

Reply:
Per #14565, we will have a way to check this unambiguously.